topology/probes/blockdev/blockdev.go

```diff
@@ -0,0 +1,515 @@
+			}
+		}
+
+		topology.AddOwnershipLink(p.Ctx.Graph, groupNode, node, nil)
+
+		// Link physical disks and DVD/rom devices to the blockdevGroupName
+		if blockdev.Type == "disk" || blockdev.Type == "rom" {
+			topology.AddLink(p.Ctx.Graph, p.Groups[blockdevGroupName], node, "connected", nil)
+		}
+
+	}
+
+	for i := range blockdev.Children {
+		childNode := p.Ctx.Graph.LookupFirstNode(graph.Metadata{"Name": p.getName(blockdev.Children[i])})
+		if childNode != nil {
+			topology.AddLink(p.Ctx.Graph, childNode, node, "connected", nil)
+		}
+	}
+
+	p.blockdevMap[blockdev.Name] = blockdevInfo{
+		ID:   blockdev.Name,
+		Node: node,
+	}
+	return groupNode
+}
+
+func (p *ProbeHandler) unregisterBlockdev(id string) {
+	p.Lock()
+	defer p.Unlock()
+
+	info, ok := p.blockdevMap[id]
+	if !ok {
+		return
+	}
+
+	p.Ctx.Graph.Lock()
+	if err := p.Ctx.Graph.DelNode(info.Node); err != nil {
+		p.Ctx.Graph.Unlock()
+		p.Ctx.Logger.Error(err)
+		return
+	}
+	p.Ctx.Graph.Unlock()
+
+	delete(p.blockdevMap, id)
+}
+
+func (p *ProbeHandler) connect() error {
+	var (
+		cmdOut []byte
+		err    error
+		result Devices
+	)
+
+	testFile := os.Getenv("SKYDIVE_BLOCKDEV_TEST_FILE")
+
+	if testFile != "" {
+		if cmdOut, err = exec.Command("cat", testFile).Output(); err != nil {
+			p.Ctx.Logger.Error(err)
+			os.Exit(1)
+		}
+	} else {
+		if cmdOut, err = exec.Command("lsblk", "-pO", "--json").Output(); err != nil {
+			p.Ctx.Logger.Error(err)
+			os.Exit(1)
+		}
+	}
+
+	if err = json.Unmarshal([]byte(cmdOut[:]), &result); err != nil {
+		p.Ctx.Logger.Error(err)
+		os.Exit(1)
+	}
+
+	// loop through the devices in the current map to make sure they haven't
+	// been removed
+	for _, current := range p.blockdevMap {
+		p.deleteIfRemoved(current, result.Blockdevices)
+	}
+
+	for i := range result.Blockdevices {
+		p.registerBlockdev(result.Blockdevices[i], "")
+	}
+
+	return nil
+}
+
+// Do adds a group for block devices, then issues a lsblk command and parsers the
+// JSON output as a basis for the blockdev links.
+func (p *ProbeHandler) Do(ctx context.Context, wg *sync.WaitGroup) error {
+	p.addGroupByName(blockdevGroupName, "")
+
+	if _, err := topology.AddLink(p.Ctx.Graph, p.Ctx.RootNode, p.Groups[blockdevGroupName], "connected", nil); err != nil {
+		p.Ctx.Logger.Error(err)
+		return err
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		after := time.After(10 * time.Minute)
+		for {
+			if err := p.connect(); err != nil {
+				p.Ctx.Logger.Error(err)
+				return
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case <-after:
+			}
+		}
+	}()
+	return nil
+}
+
+// NewProbe initializes a new topology blockdev probe
+func NewProbe(ctx tp.Context, bundle *probe.Bundle) (probe.Handler, error) {
+
+	p := &ProbeHandler{
+		blockdevMap: make(map[string]blockdevInfo),
+		Groups:      make(map[string]*graph.Node),
+		Ctx:         ctx,
+	}
+
+	p.Ctx = ctx
+	p.blockdevMap = make(map[string]blockdevInfo)
```

Comment:
what about putting this in the `p` initialization above :
```
    blockdevMap: make(map[string]blockdevInfo),
    Groups: make(map[string]*graph.Node),
```

Reply:
thanks - that's better